Implement double space full stop, switch caps lock from long click to double click

kfli716 opened the pull request.

graphemecluster commented on app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java:
```diff
@@ -1091,7 +1097,12 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
             "\t<TrimeInput>\tdetectAndSendKey()\tModifierKey, key.getEvent, KeyLabel=%s",
             key.getLabel());
-        setModifier(key);
+        if (isDoubleClick) {
```
For safety
```suggestion
        if (key.isShift() && isDoubleClick) {
```

graphemecluster commented on app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java:
```diff
@@ -1104,10 +1115,47 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
+              if (isDoubleClick) {
+                // The isHalfShapePunct is used to check whether ". " or "。" should be entered after double space.
+                mKeyboardActionListener.onDoubleSpace(
+                        lastSpaceWasConfirm ? 0 : 1,
+                        mKeyboard.isHalfShapePunct() ? ". " : "。"
```
Subtle but
```suggestion
                    lastSpaceWasConfirm ? 0 : 1,
                    mKeyboard.isHalfShapePunct() ? ". " : "。"
```

graphemecluster replied:
I like these details 👍🏻

graphemecluster commented on app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java:
```diff
@@ -1104,10 +1115,47 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
+             to delete the trailing space. And this is handled in the onDoubleSpace function. The boolean lastSpaceWasConfirm
+             is used to track the label of the last space key pressed.
+           */
+          switch (key.getLabel()) {
```
Use `Rime.isComposing()` instead – the key labels may be localized later

graphemecluster commented on app/src/main/java/hk/eduhk/typeduck/ime/text/TextInputManager.kt:
```diff
@@ -563,4 +563,9 @@ class TextInputManager private constructor() :
         trime.showDialogAboveInputView(builder.create())
     }
     */
+
+    // added to handle doubleSpaceFullStop
+    override fun onDoubleSpace(charToDel: Int, punct: CharSequence) {
+        activeEditorInstance.doubleSpaceEvent(charToDel, punct)
+    }
```
I prefer moving this above `showOptionsDialog`

graphemecluster reviewed (commented):
- Please remove [KeyboardView.java#L1480-L1487](https://github.com/TypeDuck-HK/TypeDuck-Android/blob/84ef8d49b93678d2199c28e9730c82e3e7f2bb88/app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java#L1480-L1487)
- Could you rebase your change and help me uncomment the option in [app/src/main/res/xml/preference.xml](https://github.com/TypeDuck-HK/TypeDuck-Android/blob/50ea4cb6c0f3685acce9d957e1582ca5e4843eaf/app/src/main/res/xml/preference.xml#L54-L61)?

graphemecluster commented:
Oh, but the behavior does not match what Gboard and iOS built-in keyboard currently do – the click when composing should not be counted towards double click. That means we should only detect double click on spacebar when `Rime.isComposing()` is false and to insert a full stop during composition a user needs to click the spacebar a single time for confirmation to exit the composition mode before a double click.

Plus, I plan to add and option to change the behavior of the spacebar when composing (flip page, insert candidate, move to next candidate, keep input), so spacebar double clicking should be disabled during composition to prevent any disruption.

graphemecluster commented on app/src/main/java/hk/eduhk/typeduck/ime/keyboard/KeyboardView.java:
```diff
@@ -1091,7 +1097,12 @@ private void detectAndSendKey(int index, int x, int y, long eventTime, KeyEventT
-        setModifier(key);
+        if (isDoubleClick) {
+          mKeyboard.setShifted(true, true);
+          //isDoubleClick = false;
```
Do you think this line is needed? Let’s remove it if not

graphemecluster replied:
Do you think this line is needed? Let’s remove it if not

graphemecluster commented:
But remember, shift double clicking should still work as usual during composition

kfli716 commented:
Changes:
1. Changed the code accoding to the suggestions.
2. Added a function isSpace() to Key.java to distinguish if the pressed key is the space key, without using key label.
3. Changed the behavior to be more aligned with the GBoard. Now the doubleSpaceToFullStop function will not trigger if the last character is a space key or in " ！？、，。!?.,". And if it triggers, the full stop generated is determined by the input language being used.

graphemecluster commented:
Done? If so, I’m going to investigate this today

graphemecluster commented:
Never mind, tomorrow is fine

kfli716 commented:
Yes, and if you have time, please also review the PR about launch issue (the CMake file) and let me know whether there is any problem, thank you very much! 🙏
